v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/build-system.ts

```diff
@@ -131,10 +177,27 @@ export class SolidityBuildSystemImplementation implements SolidityBuildSystem {
       },
     );
 
-    const isSuccessfulBuild = results.every(
-      (result) => !this.#hasCompilationErrors(result),
+    const uncachedResults = results.filter((result) => !result.cached);
+    const uncachedSuccessfulResults = uncachedResults.filter(
+      (result) => !this.#hasCompilationErrors(result.compilerOutput),
     );
 
+    // NOTE: We're not waiting for the writes and clean to finish because we
+    // will only care about the result of these operations in subsequent runs
+    void Promise.all(
```

Comment:
I think this is unsafe, which we would be ok though if we implement this change: https://github.com/NomicFoundation/hardhat/pull/6129/files#r1944738230

I think I'd move it to its own private method (a sync `#updateCache`?), and better explain why it's safe.

Reply:
It was! You're right! I did move it to a separate method and added a longer docstring with a callout that it might have to be reconsidered in the future. Great catch, thank you 🙇